### ISSUE TITLE
Tips for Debugging a Language Parser

### ISSUE THREAD

**Issue by simplygreatwork** [francisrstokes/arcsecond#10]
Could you offer any tips for debugging? Whenever I'm coding imperatively, I often print to the console for debugging. But that's often not easy to do even with tapParser.

I'm trying to make a block language parser with support for parentheses, newline separation, and semicolon separation. I began with the JSON parser example. But my parser is more strict than I would like. Could you offer any pointers so that the last code block in my example below will parse fine (the block inside if false)

https://gist.github.com/simplygreatwork/a29e5b7391ea438b75e87461d4a6c8c2

Also, is it possible to separate using a choice of either newline or semicolon as a single parser? I was not yet able to get that it working either.

**Comment by francisrstokes** [comment]
Hi @simplygreatwork ,

I can sympathsise with the debugging difficulties - I'm solving this in the next version of arcsecond which will be release soon. Until then, you can gain back some of the debugging capabilities with a bit of help from generators:

```javascript
const {Parser} = require('arcsecond');

const coroutine = g => Parser.of().chain(_ => {
    const generator = g();

    const step = nextValue => {
      const result = generator.next(nextValue);
      const value = result.value;
      const done = result.done;

      if (!done && (!value || typeof value.chain !== 'function')) {
        throw new Error(`[coroutine] yielded values must be Parsers, got ${result.value}.`);
      }

      return done ? Parser.of(value) : value.chain(step);
    };

    return step();
  });
};
```

Which can be used like:

```javascript
const someParser = coroutine(function* () {
   const firstGetSomeLetters = yield letters;
   console.log(`Result: ${firstGetSomeLetters}`);

   const thenGetSomeDigits = yield digits;
   console.log(`Digits were: ${thenGetSomeDigits}`);

   // Return whatever you want as the parser result
   return { first: firstGetSomeLetters, second: thenGetSomeDigits };
});

parse(someParser)('hello42');
// [console.log] Result: hello
// [console.log] Digits were: 42
// -> { first: 'hello', second: '42' }
```

If you're used to the debugger, you can actually put breakpoints inside the the generator function and step through.

`coroutine` will be in the library by default in 2.0.0.

Hope this helps!

**Comment by francisrstokes** [comment]
I released version 2.0.0 this morning, which I believe addressed many of the difficulties with debugging, as well as adding an entire tutorial section about different approaches.

I'll close this issue for now, but if you come across an other difficulties while using the new major version, feel free to open another.
